usine.py

```diff
@@ -1,23 +1,345 @@
+        hostport = user_hostport.pop()
+        user = user_hostport[0] if user_hostport and user_hostport[0] else None
+
+        # IPv6: can't reliably tell where addr ends and port begins, so don't
+        # try (and don't bother adding special syntax either, user should avoid
+        # this situation by using port=).
+        if hostport.count(':') > 1:
+            host = hostport
+            port = None
+        # IPv4: can split on ':' reliably.
+        else:
+            host_port = hostport.rsplit(':', 1)
+            host = host_port.pop(0) or None
+            port = host_port[0] if host_port and host_port[0] else None
+
+        if port is not None:
+            port = int(port)
+
+        return {'username': user, 'hostname': host, 'port': port}
+
+    def execute(self, cmd, **kwargs):
+        channel = self._transport.open_session()
+        prefix = self.prefix
+        if self.cd:
+            cmd = f'cd {self.cd}; {cmd}'
+        if self.env:
+            env = ' '.join(f'{k}={v}' for k, v in self.env.items())
+            prefix = f'{prefix} {env}'
+        cmd = self.format(prefix + " sh -c '" + cmd + "'")
+        print(gray(cmd))
+        channel.exec_command(cmd)
+        stdout = channel.makefile('r', -1)
+        stderr = channel.makefile_stderr('r', -1)
+        buf = b''
+        while True:
+            data = stdout.read(1)
+            if not data:
+                if buf:
+                    sys.stdout.write(buf.decode())
+                break
+            buf += data
+            if buf.endswith(b'\n'):
+                sys.stdout.write(buf.decode())
+                buf = b''
+        ret = Status(stdout, stderr, channel.recv_exit_status())
+        channel.close()
+        if ret.code:
+            red(ret.stderr)
+            sys.exit(ret.code)
+        return ret
+
+    def format(self, tpl):
+        return self.formatter.vformat(tpl, None, self.context)
+
+    @property
+    def sftp(self):
+        if not self._sftp:
+            self._sftp = self._client.open_sftp()
+        return self._sftp
+
+
+@contextmanager
+def connect(*args, **kwargs):
+    global client
+    client = Client(*args, **kwargs)
+    yield client
+    client.close()
 
 
 def run(cmd):
-    if connection.context['sudo']:
-        cmd = f'sudo {cmd}'
-    connection.execute(cmd)
+    return client.execute(cmd)
+
+
+def exists(path):
+    return bool(run(f'if [ -e "{path}" ]; then echo 1; fi').stdout)
+
+
+@formattable
+def mkdir(path, parents=True, mode=None):
+    return run('mkdir {parents:B} {mode:V} {path}')
+
+
+@formattable
+def chown(mode, path, recursive=True, preserve_root=True):
+    return run('chown {recursive:B} {mode} {path}')
+
+
+@formattable
+def ls(path, all=True, human_readable=True, size=True, list=True):
+    return run('ls {all:B} {human_readable:B} {size:B} {list:S} {path}')
+
+
+def mv(src, dest):
+    return run(f'mv {src} {dest}')
+
+
+def put(local, remote, owner=None):
```

Comment:
Can we use the ssh client instead, so we can take advantage of `with sudo/with cd` and such?